core/services/pg/q.go

```diff
@@ -68,6 +68,17 @@ func WithParentCtx(ctx context.Context) func(q *Q) {
 	}
 }
 
+// If the parent has a timeout, just use that instead of DefaultTimeout
+func WithParentCtxInheritTimeout(ctx context.Context) func(q *Q) {
+	return func(q *Q) {
+		q.ParentCtx = ctx
+		deadline, ok := q.ParentCtx.Deadline()
+		if ok {
+			q.QueryTimeout = time.Until(deadline) + 1*time.Second
```

Comment:
Why pad one second here?

Reply:
I just figured it seems slightly less risky (in terms of potential race conditions) if the parent and child time out in a deterministic order rather than in random order.  Ideally, there would be an "infinite" value it could be set to, but anything strictly grater than the parent's timeout makes it clear that it will wait for the parent to time out and not earlier, which is the intention.

Reply:
(Leaving it up to the parent was suggested by @samsondav in an earlier PR comment.)

Reply:
This is not quite what I had in mind. I was imagining more like a `WithReplacementTimeout` or something that would accept a `time.Duration` and always override the default.

Reply:
I'm not sure I follow the reasoning here. We can't actually control anything to do with the ordering/timing, and shouldn't be depending on any assumptions here - they can and will be violated by arbitrary scheduler reordering when resources are limited, no matter how much time we pad with.

Separately, if we are trying to avoid setting `q.QueryTimeout == 0`, then don't we need to inspect the result of `time.Until(deadline)` which could be <=0? What about something like this:
```suggestion
			q.QueryTimeout = time.Until(deadline)
			if q.QueryTimeout <= 0 {
				q.QueryTimeout = 1 * time.Nanosecond
			}
```

Reply:
I think I didn't explain it well.  It doesn't depend on any assumptions about the order, it's just that I felt a bit nervous about setting both of the deadlines to exactly the same moment.  For the same reason we add a "jitter" factor in many places in our code, I figured it might be better to offset them.  You're probably right that it's unnecessary, especially if there's already some healthy jitter in there from the scheduler.  Happy to remove the 1s offset.

Regarding the possibility that the deadline has already expired, I don't see any reason to worry about that.
For the present use case, the caller of this function creates the context with a 1 hour (default) deadline and then immediately passes it to here, so even if there were an enormous resource crunch, there's no way it would take an hour to push a few registers on the stack and enter the function.  In the more general case, considering all possible potential future uses of this function, I suppose it's possible a caller might pass an expired context in, but I don't think that would cause any (additional) problem.  (I say "additional", since I assume if this happened it would mean that queries are timing out before they're even getting submitted to the db... which would indicate a big resource/slowness problem with the node.)

Maybe it will help to look at the code that this is a workaround for (`q.Context()`):
https://github.com/smartcontractkit/chainlink/blob/develop/core/services/pg/q.go#L140-L153

`q.Context()` is called at the beginning of each of the query wrappers exposed in `services/pg` (Get, Select, Exec, etc.) to create a new context based on this context.  It checks for `q.QueryTimeout > 0` to see if it should replace the current timeout with the default timeout.  So if it's negative, it will just use the default timeout, same as if it were set to 0.  I'm not sure what happens if you try to create a context with an expired parent.  I guess ideally it would mark the child as immediately cancelled/done, but even if it doesn't do that... using the default timeout (1s) instead of immediately returning seems harmless enough.

Reply:
I don't think the same reasoning for using jitter on scheduled work or external resources applies here, both because we don't force the runtime to do anything (it will get to things in it's own time) and because these timeouts are generally _realeasing_ resources by abandoning ongoing work. So I would vote to remove the extra padding - I think it will just confuse future readers.

Is this still relevant though with @samsondav comment?